### PR TITLE
Migrate ast lowering to session diagnostic

### PR DESCRIPTION
**Pull request by JeanCASPAR**
I migrated the whole rustc_ast_lowering crate to session diagnostic *except* the for the use of `span_fatal` at /compiler/rustc_ast_lowering/src/expr.rs#L1268 because `#[fatal(...)]` is not yet supported (see https://github.com/rust-lang/rust/pull/100694).

### CODE REVIEW

File: compiler/rustc_ast_lowering/src/errors.rs
Comment (davidtwco) [inline compiler/rustc_ast_lowering/src/errors.rs:199]: nit: it would be nice if these names were slightly more descriptive

File: compiler/rustc_ast_lowering/src/errors.rs
Comment (davidtwco) [inline compiler/rustc_ast_lowering/src/errors.rs:209]: nit: it would be nice if these names were slightly more descriptive

**Review by davidtwco — CHANGES_REQUESTED** [review]
This is really great, thanks for working on it. I've left a few comments.

File: compiler/rustc_ast_lowering/src/lib.rs
Comment (JeanCASPAR) [reply compiler/rustc_ast_lowering/src/lib.rs:39]: Yes it is.

File: compiler/rustc_ast_lowering/src/lib.rs
Comment (JeanCASPAR) [reply compiler/rustc_ast_lowering/src/lib.rs]: I used `DiagnosticArgFromDisplay` instead.

**Review by davidtwco — APPROVED** [review]

File: compiler/rustc_ast_lowering/src/errors.rs
Comment (JeanCASPAR) [reply compiler/rustc_ast_lowering/src/errors.rs]: Done

File: compiler/rustc_ast_lowering/src/errors.rs
Comment (JeanCASPAR) [reply compiler/rustc_ast_lowering/src/errors.rs]: Done

**Comment by davidtwco** [comment]
Could you remove the merge commits? We have a [no merge policy](https://rustc-dev-guide.rust-lang.org/git.html#no-merge-policy) - thanks!

**Comment by davidtwco** [comment]
r=me after merge commits removed

**Comment by JeanCASPAR** [comment]
Don't know if this intended, @davidtwco ?

**Comment by davidtwco** [comment]
> Don't know if this intended, @davidtwco ?

It is, only approved reviewers can r+, I just left that comment in case another reviewer looked at the PR and saw that they could approve on my behalf after you resolved the comment.

@bors r+

**Pull request merged**
